pkg/utils/ldap/identity.go

```diff
@@ -0,0 +1,574 @@
+	filter := fmt.Sprintf("(objectclass=%s)", i.User.Objectclass)
+	if i.User.Filter != "" {
+		filter = fmt.Sprintf("(&%s%s)", i.User.Filter, filter)
+	}
+	searchRequest := ldap.NewSearchRequest(
+		dn, i.User.scopeVal, ldap.NeverDerefAliases, 1, 0, false,
+		filter,
+		[]string{
+			i.User.Schema.DisplayName,
+			i.User.Schema.ID,
+			i.User.Schema.Mail,
+			i.User.Schema.Username,
+		},
+		nil,
+	)
+	log.Debug().Str("backend", "ldap").Str("basedn", dn).Str("filter", filter).Int("scope", i.User.scopeVal).Msg("LDAP Search")
+	res, err := lc.Search(searchRequest)
+
+	if err != nil {
+		log.Debug().Str("backend", "ldap").Err(err).Str("dn", dn).Msg("Error looking up user by DN")
+		return nil, errtypes.NotFound(dn)
+	}
+	if len(res.Entries) == 0 {
+		return nil, errtypes.NotFound(dn)
+	}
+
+	return res.Entries[0], nil
+}
+
+// GetLDAPUsers searches for users using a prefix-substring match on the user
+// attributes. Returns a slice of matching ldap.Entries
+func (i *Identity) GetLDAPUsers(log *zerolog.Logger, lc ldap.Client, query string) ([]*ldap.Entry, error) {
+	filter := i.getUserFindFilter(query)
+	searchRequest := ldap.NewSearchRequest(
+		i.User.BaseDN,
+		i.User.scopeVal, ldap.NeverDerefAliases, 0, 0, false,
+		filter,
+		[]string{i.User.Schema.ID, i.User.Schema.Username, i.User.Schema.Mail, i.User.Schema.DisplayName, i.User.Schema.UIDNumber, i.User.Schema.GIDNumber},
+		nil,
+	)
+
+	log.Debug().Str("backend", "ldap").Str("basedn", i.User.BaseDN).Str("filter", filter).Int("scope", i.User.scopeVal).Msg("LDAP Search")
+	sr, err := lc.Search(searchRequest)
+	if err != nil {
+		log.Debug().Str("backend", "ldap").Err(err).Str("filter", filter).Msg("Error searching users")
+		return nil, errtypes.NotFound(query)
+	}
+	return sr.Entries, nil
+}
+
+// GetLDAPUserGroups looks up the group member ship of the supplied LDAP user entry.
+// Returns a slice of strings with groupids
+func (i *Identity) GetLDAPUserGroups(log *zerolog.Logger, lc ldap.Client, userEntry *ldap.Entry) ([]string, error) {
+	var memberValue string
+
+	if strings.ToLower(i.Group.Objectclass) == "posixgroup" {
+		// posixGroup usually means that the member attribute just contains the username
+		memberValue = userEntry.GetEqualFoldAttributeValue(i.User.Schema.Username)
+	} else {
+		// In all other case we assume the member Attribute to contain full LDAP DNs
+		memberValue = userEntry.DN
+	}
+
+	filter := i.getGroupMemberFilter(memberValue)
+	searchRequest := ldap.NewSearchRequest(
+		i.Group.BaseDN, i.Group.scopeVal,
+		ldap.NeverDerefAliases, 0, 0, false,
+		filter,
+		[]string{i.Group.Schema.ID},
+		nil,
+	)
+
+	log.Debug().Str("backend", "ldap").Str("basedn", i.Group.BaseDN).Str("filter", filter).Int("scope", i.Group.scopeVal).Msg("LDAP Search")
+	sr, err := lc.Search(searchRequest)
+	if err != nil {
+		log.Debug().Str("backend", "ldap").Err(err).Str("filter", filter).Msg("Error looking up group memberships")
+		return []string{}, err
+	}
+
+	groups := []string{}
```

Comment:
Better to use make for allocating slices, especially when the capacity is know: https://github.com/uber-go/guide/blob/master/style.md#specifying-slice-capacity